File: server/achievements/evaluate.go
package achievements

type Observation struct {
	Facts      map[string]bool
	Counters   map[string]int64
	ExitCount  int64
	Generators map[string]int64
}

func Eligible(condition Condition, observation Observation) bool {
	switch condition.Kind {
	case ConditionFactPresent:
		return observation.Facts[condition.FactKind]
	case ConditionCounterAtLeast:
		return observation.Counters[condition.Counter] >= condition.Minimum
	case ConditionExitCountAtLeast:
		return observation.ExitCount >= condition.Minimum
	case ConditionOwnsGeneratorAtLeast:
		return observation.Generators[condition.GeneratorID] >= condition.Minimum
	case ConditionAllOf:
		for _, child := range condition.Children {
			if !Eligible(child, observation) {
				return false
			}
		}
		return true
	default:
		return false
	}
}

// NewlyEarned evaluates every definition against one pre-achievement snapshot.
// Returned definitions retain catalog byte order and never include a lifetime latch.
func (catalog *Catalog) NewlyEarned(runEarned, lifetimeEarned map[string]bool, run, career Observation) ([]Definition, error) {
	if catalog == nil || runEarned == nil || lifetimeEarned == nil {
		return nil, ErrInvalidCatalog
	}
	result := []Definition{}
	for _, definition := range catalog.Definitions {
		if runEarned[definition.ID] || lifetimeEarned[definition.ID] {
			continue
		}
		observation := run
		if definition.ConditionScope == ScopeCareer {
			observation = career
		}
		if Eligible(definition.Condition, observation) {
			result = append(result, cloneDefinition(definition))
		}
	}
	return result, nil
}

File: client/src/achievements/evaluate.ts
import type { AchievementCatalog, AchievementCondition, AchievementDefinition } from "./catalog";
import { MAX_EXACT_INTEGER } from "./catalog";

export interface AchievementObservation {
  readonly facts: ReadonlySet<string>;
  readonly counters: Readonly<Record<string, number>>;
  readonly exitCount: number;
  readonly generators: Readonly<Record<string, number>>;
}

export function achievementEligible(condition: AchievementCondition, observation: AchievementObservation): boolean {
  switch (condition.kind) {
    case "fact_present": return observation.facts.has(condition.factKind);
    case "counter_at_least": return (observation.counters[condition.counter] ?? 0) >= condition.minimum;
    case "exit_count_at_least": return observation.exitCount >= condition.count;
    case "owns_generator_at_least": return (observation.generators[condition.generatorId] ?? 0) >= condition.count;
    case "all_of": return condition.conditions.every((child) => achievementEligible(child, observation));
  }
}

export function newlyEarned(catalog: AchievementCatalog, runEarned: ReadonlySet<string>, lifetimeEarned: ReadonlySet<string>, run: AchievementObservation, career: AchievementObservation): readonly AchievementDefinition[] {
  return Object.freeze(catalog.definitions.filter((definition) => !runEarned.has(definition.id) && !lifetimeEarned.has(definition.id) && achievementEligible(definition.condition, definition.conditionScope === "run" ? run : career)));
}

export function achievementScore(catalog: AchievementCatalog, earned: ReadonlySet<string>): number {
  let result = 0;
  for (const id of earned) {
    const definition = catalog.byId.get(id);
    if (!definition || result > MAX_EXACT_INTEGER - definition.scoreGrant) throw new RangeError("invalid earned achievement set");
    result += definition.scoreGrant;
  }
  return result;
}
